crates/bootstrap_srv/src/server.rs

```diff
@@ -0,0 +1,333 @@
+            method: req.method().as_str().to_string(),
+            path,
+            req,
+        };
+
+        handler.handle()?;
+    }
+    Ok(())
+}
+
+struct Handler<'lt> {
+    config: &'lt Config,
+    store: &'lt crate::Store,
+    space_map: &'lt crate::SpaceMap,
+    method: String,
+    path: Vec<String>,
+    req: tiny_http::Request,
+}
+
+impl<'lt> Handler<'lt> {
+    /// Wrap the handle call so we can respond to the client with errors.
+    pub fn handle(mut self) -> std::io::Result<()> {
+        match self.handle_inner() {
+            Ok((status, body)) => self.respond(status, body),
+            Err(err) => self.respond(
+                500,
+                serde_json::to_string(&serde_json::json!({
+                    "error": format!("{err:?}"),
+                }))?
+                .into_bytes(),
+            ),
+        }
+    }
+
+    /// Dispatch to the correct handlers.
+    fn handle_inner(&mut self) -> std::io::Result<(u16, Vec<u8>)> {
+        if let Some(cmd) = self.path.pop() {
+            match (self.method.as_str(), cmd.as_str()) {
+                ("GET", "health") => {
+                    return Ok((200, b"{}".to_vec()));
+                }
+                ("GET", "bootstrap") => {
+                    return self.handle_boot_get();
+                }
+                ("PUT", "bootstrap") => {
+                    return self.handle_boot_put();
+                }
+                _ => (),
+            }
+        }
+        Ok((400, b"{\"error\":\"bad request\"}".to_vec()))
+    }
+
+    /// Respond to a request for the agent infos within a space.
+    fn handle_boot_get(&mut self) -> std::io::Result<(u16, Vec<u8>)> {
+        let space = self.path_to_bytes()?;
+
+        let res = self.space_map.read(&space)?;
+
+        Ok((200, res))
+    }
+
+    /// Validate an incoming agent info and put it in the store if appropriate.
+    fn handle_boot_put(&mut self) -> std::io::Result<(u16, Vec<u8>)> {
+        use ed25519_dalek::*;
+
+        let now = crate::now();
+
+        let space = self.path_to_bytes()?;
+        let agent = self.path_to_bytes()?;
+
+        let info_raw = self.read_body()?;
+        let info = crate::ParsedEntry::from_slice(&info_raw)?;
+
+        // validate agent matches url path
+        if *agent != *info.agent.as_bytes() {
+            return Err(std::io::Error::other("InvalidAgent"));
+        }
+
+        // validate space matches url path
+        if space != info.space {
+            return Err(std::io::Error::other("InvalidSpace"));
+        }
+
+        // validate created at is not older than 3 min ago
+        if info.created_at
+            < now - (std::time::Duration::from_secs(60 * 3).as_micros() as i64)
+        {
+            return Err(std::io::Error::other("InvalidCreatedAt"));
+        }
+
+        // validate created at is less than 3 min in the future
+        if info.created_at
+            > now + (std::time::Duration::from_secs(60 * 3).as_micros() as i64)
```

Comment:
Maybe we should move this value to a const in the config file.

Reply:
Done: https://github.com/holochain/kitsune2/pull/10/commits/263d5ab3f9e742b19c46ce7a11d370205ad904ae